### PR TITLE
Simplify Unproven struct

### PR DESCRIPTION
**Pull request by ConstanceBeguier**
We would like to remove the generic circuit parameter from Unproven structure.
To do that, we create a Witnesses structure which contains the witnesses of the circuit but not the circuit type (OrchardZSA or OrchardVanilla).

Key changes:
    - Updated `src/builder.rs` to replace `Circuit` with `Witnesses` in several function signatures and type definitions.
    - Adjusted `src/circuit.rs` to define and use the new `Witnesses` structure instead of `Circuit`.

### CODE REVIEW

File: benches/circuit.rs [hunk 1]
```diff
@@ -68,7 +68,7 @@ fn criterion_benchmark<FL: OrchardFlavorBench>(c: &mut Criterion) {
                 b.iter(|| {
                     bundle
                         .authorization()
-                        .create_proof(&pk, &instances, rng)
+                        .create_proof::<FL>(&pk, &instances, rng)
```
Comment (PaulLaux) [inline benches/circuit.rs:71]: nice catch

File: src/builder.rs [hunk 2]
```diff
@@ -935,23 +935,32 @@ impl<P: fmt::Debug, S: InProgressSignatures> Authorization for InProgress<P, S>
...
+        let circuits: Vec<Circuit<C>> = self
+            .proof
+            .witnesses
+            .iter()
+            .map(|witnesses| Circuit::<C> {
+                witnesses: witnesses.clone(),
+                phantom: std::marker::PhantomData,
+            })
+            .collect();
+        Proof::create(pk, &circuits, instances, rng)
```
Comment (PaulLaux) [inline src/builder.rs:959]: Open question: 
Can we add some kind of protection from triggering ZSA proof generation using Vanilla witness or vice versa?
what is the cost?

File: src/circuit/circuit_vanilla.rs [hunk 3]
```diff
@@ -858,28 +861,30 @@ mod tests {
...
+                rivk: Value::unknown(),
+                g_d_new: Value::unknown(),
+                pk_d_new: Value::unknown(),
+                v_new: Value::unknown(),
+                psi_new: Value::unknown(),
+                rcm_new: Value::unknown(),
+                rcv: Value::unknown(),
+                asset: Value::unknown(),
+                split_flag: Value::unknown(),
+            },
```
Comment (PaulLaux) [inline src/circuit/circuit_vanilla.rs:887]: This should be replaced with
`witnesses: Witnesses::default()` or similar (all occurrences)
(need to make sure `default()` properly defined.)

File: src/builder.rs (hunk 2 above)
Comment (ConstanceBeguier) [reply src/builder.rs]: Added circuit_flavor into Unproven struct

File: src/orchard_flavor.rs [hunk 4]
```diff
@@ -10,8 +10,28 @@ pub struct OrchardVanilla;
...
+impl OrchardFlavor for OrchardVanilla {
+    fn flavor() -> Flavor {
+        Flavor::OrchardVanillaFlavor
+    }
+}
+impl OrchardFlavor for OrchardZSA {
+    fn flavor() -> Flavor {
+        Flavor::OrchardZSAFlavor
+    }
+}
```
Comment (PaulLaux) [inline src/orchard_flavor.rs:37]: Function is nice but `CONST` is even simpler:
```rust
/// A trait binding the common functionality between different Orchard protocol flavors.
pub trait OrchardFlavor: OrchardDomainCommon + OrchardCircuit + OrchardHash {
    const FLAVOR : Flavor;
}

impl OrchardFlavor for OrchardVanilla {
    const FLAVOR : Flavor = Flavor::OrchardVanillaFlavor;
}
impl OrchardFlavor for OrchardZSA {
    const FLAVOR : Flavor = Flavor::OrchardZSAFlavor;
}

/// in builder.rs
proof: Unproven {
        witnesses,
        circuit_flavor: FL::FLAVOR,
    },
sigs: Unauthorized { bsk },
```
Everything else inducing the enum is unchanged.

File: src/orchard_flavor.rs [hunk 5]
```diff
@@ -10,8 +10,28 @@ pub struct OrchardVanilla;
 #[derive(Debug, Clone, Default)]
 pub struct OrchardZSA;
 
+/// Represents the flavor of the Orchard protocol.
```
Comment (PaulLaux) [inline src/orchard_flavor.rs:13]: please add
`/// This enables conditional execution during runtime based on the flavor of the Orchard protocol.`

**Review by PaulLaux — APPROVED** [review]
added minor comments

File: src/orchard_flavor.rs (hunk 4 above)
Comment (ConstanceBeguier) [reply src/orchard_flavor.rs:35]: Done

File: src/orchard_flavor.rs (hunk 5 above)
Comment (ConstanceBeguier) [reply src/orchard_flavor.rs:13]: Done

**Pull request merged**